app/src/main/java/org/mozilla/fenix/settings/SettingsFragment.kt

```diff
@@ -427,24 +429,62 @@ class SettingsFragment : PreferenceFragmentCompat() {
     }
 
     private fun getClickListenerForMakeDefaultBrowser(): Preference.OnPreferenceClickListener {
-        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
-            Preference.OnPreferenceClickListener {
-                val intent = Intent(android.provider.Settings.ACTION_MANAGE_DEFAULT_APPS_SETTINGS)
-                startActivity(intent)
-                true
+        return when {
+            Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q -> {
+                Preference.OnPreferenceClickListener {
+                    requireContext().getSystemService(RoleManager::class.java).also {
+                        if (!it.isRoleHeld(RoleManager.ROLE_BROWSER)) {
+                            startActivityForResult(it.createRequestRoleIntent(RoleManager.ROLE_BROWSER), 0)
+                        } else {
+                            navigateUserToDefaultAppsSettings()
+                        }
+                    }
+                    true
+                }
             }
-        } else {
-            Preference.OnPreferenceClickListener {
-                (activity as HomeActivity).openToBrowserAndLoad(
-                    searchTermOrURL = SupportUtils.getSumoURLForTopic(
-                        requireContext(),
-                        SupportUtils.SumoTopic.SET_AS_DEFAULT_BROWSER
-                    ),
-                    newTab = true,
-                    from = BrowserDirection.FromSettings
-                )
-                true
+            Build.VERSION.SDK_INT >= Build.VERSION_CODES.N -> {
+                Preference.OnPreferenceClickListener {
+                    navigateUserToDefaultAppsSettings()
+                    true
+                }
             }
+            else -> {
+                Preference.OnPreferenceClickListener {
+                    (activity as HomeActivity).openToBrowserAndLoad(
+                            searchTermOrURL = SupportUtils.getSumoURLForTopic(
+                                    requireContext(),
+                                    SupportUtils.SumoTopic.SET_AS_DEFAULT_BROWSER
+                            ),
+                            newTab = true,
+                            from = BrowserDirection.FromSettings
+                    )
+                    true
+                }
+            }
+        }
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        /*
+        If role manager doesn't show in-app browser changing dialog for a reason, navigate user to
+        Default Apps Settings.
+
+        Possible reasons:
+        1- There is no such role in Android system itself
+        2- User selected "don't show again" in role manager dialog before
+        3- User not selected firefox in dialog
+         */
```

Comment:
I like this comment, but I think we can just keep the first sentence and take out "Possible reasons"